### DISCUSSION TITLE
How to build cache with MicroStream?

### DISCUSSION THREAD

**Discussion by nhthong2001** [microstream-one/microstream#519] [Q&A]
Hello MicroStream team,
I want to build a cache that uses MicroStream to store Objects which can be Set<Pojo>, List<Pojo> or just a Pojo object. These pojos objects can be repeatedly saved at runtime. Is there a way to not have to duplicate these Pojos in MicroStream's memory?

**Comment by hg-ms** [comment]
Hello,
I’am not sure if I understood your question correctly, what you men with “…not have to duplicate these Pojos in MicroStream's memory?”
In general, items that are persisted more than once are not duplicated in memory, the internal cached data just gets updated. A duplication only happens at the persistence level (in the storage files) that will be cleaned up depending on the housekeeping configuration.
If you have an object referenced from several other ones it is also not duplicated…

**Reply by nhthong2001** [reply]
For example, Set<user> have [user(id=1), user(id=2)], List<user> have user(id=1), user(id=3), when I store List and Set into MicroStream, I don't want user(id=1) is duplicated in storage

**Reply by hg-ms** [reply]
Microstream does not duplicate objects. 
If the Set and the List reference the same “User” instance (reference equality) there is also no duplication in the storage. Microstream detects that both collections reference to the same “User” instance and therefore persists that User only once.
This is also true in any other case where an object is referenced from more the one other object.

**Reply by lucasmogari** [reply]
@hg-ms  I'm sorry to reply to this issue, but I have a related question. Does microstream do that for boxed primitives too?

For example:

```
public class User {
    private Long id;
    private Integer age;
}
```

The age field will be stored as the same instance for this list [user(id=1, age=20), user(id=2, age=20)]?

**Comment by hg-ms** [comment]
Hi,
Boxed primitives behave the same way, they are just objects that are referenced.

There is one important detail regarding boxed primitives: Some Java-APIs may return instances from an internal cache, instead of creating new instances (e.g.: java.lang.Integer.valueOf(int)).

In your example both fields of the Users can ether reference a shared object or independent objects, it depends on the objects they reference.

When working with Microstream, you’re working with java objects that reference each other, they build an object graph where the objects are nodes and references are the edges. When Microstream persists that graph objects are not duplicated, primitive values are part of the node.

**Reply by lucasmogari** [reply]
Thanks for your reply! So if I want to save memory and reference the same boxed primitive instance I would have to do some sort of deduplication by myself right?

**Comment by hg-ms** [comment]
Yes, if you need to save memory reducing the number of persisted objects is a good idea as every persisted objects also creates some management overhead in Microstream in addition to its data.
If doing so you may also consider replacing referenced primitives with real primitive values, because every persisted reference requires 8 bytes in addition to the referenced object.
